### ISSUE TITLE
Is doing a lookup thread safe?

### ISSUE THREAD

**Issue by dmarlow** [flipbit/whois#17]
I'm curious because it looks like it's getting deadlocked, but can't confirm without digging deeper.

**Comment by flipbit** [comment]
How many requests are you putting through the same instance of the library? (threads / requests/sec)
Are there any specific domains that are causing the deadlock?

I suspect the issue maybe due to network timeouts - I'll see if I can recreate and upload a fix.

**Comment by dmarlow** [comment]
I recall doing something like 25 or 50 concurrently. Once I changed it to instantiate an instance of the WhoisLookup per thread, the issue went away. There weren't any domains in particular. 

Though I did notice that some could potentially take a while and there is no timeout that I could specify. I worked around it by doing this:

```
var task = new WhoisLookup().LookupAsync(x.Domain);
if (await Task.WhenAny(task, Task.Delay(5000)) == task) 
{
    // Check task.IsFaulted to see if an exception occurred
    // or else it completed successfully
}
else 
{
    // Handle timeout
}
```

**Comment by flipbit** [comment]
I've found an issue in the parsing code, around an unprotected access to a generic `List<T>` which I've switched to using a `ConcurrentDictionary<T>`.  I've tested with 25 concurrent threads with no errors.

I'm planning a major new release of the library in the next week or so that includes support for parsing 140+ new TLDs, so I'll roll the fix up into that release.

**Comment by flipbit** [comment]
This should be fixed now in version 3.0.0
